diff --git a/hopper/setup.py b/hopper/setup.py
index 87f6f45..ecc52e0 100755
--- a/hopper/setup.py
+++ b/hopper/setup.py
@@ -22,7 +22,8 @@ import urllib.error
 from wheel.bdist_wheel import bdist_wheel as _bdist_wheel
 
 import torch
-from torch.utils.cpp_extension import BuildExtension, CppExtension, CUDAExtension, CUDA_HOME
+from torch.utils.cpp_extension import CppExtension, CUDAExtension, CUDA_HOME
+from torch.utils.cpp_extension import BuildExtension as _BuildExtension
 
 
 # with open("../README.md", "r", encoding="utf-8") as fh:
@@ -451,7 +452,7 @@ if not SKIP_CUDA_BUILD:
     # We want to use the nvcc front end from 12.6 however, since if we use nvcc 12.8
     # Cutlass 3.8 will expect the new data types in cuda.h from CTK 12.8, which we don't have.
     # For CUDA 13.0+, use system nvcc instead of downloading CUDA 12.x toolchain
-    if bare_metal_version >= Version("12.3") and bare_metal_version < Version("13.0") and bare_metal_version != Version("12.8"):
+    if not IS_WINDOWS and bare_metal_version >= Version("12.3") and bare_metal_version < Version("13.0") and bare_metal_version != Version("12.8"):
         download_and_copy(
             name="nvcc",
             src_func=lambda system, arch, version: f"cuda_nvcc-{system}-{arch}-{version}-archive/bin",
@@ -574,12 +575,12 @@ if not SKIP_CUDA_BUILD:
     if DISABLE_BACKWARD:
         sources_bwd_sm90 = []
         sources_bwd_sm80 = []
-    
+
     # Choose between flash_api.cpp and flash_api_stable.cpp based on torch version
     torch_version = parse(torch.__version__)
     target_version = parse("2.9.0.dev20250830")
     stable_args = []
-      
+
     if torch_version >= target_version:
         flash_api_source = "flash_api_stable.cpp"
         stable_args = ["-DTORCH_TARGET_VERSION=0x0209000000000000"]  # Targets minimum runtime version torch 2.9.0
@@ -703,6 +704,89 @@ class CachedWheelsCommand(_bdist_wheel):
             # If the wheel could not be downloaded, build from source
             super().run()
 
+class BuildExtension(_BuildExtension):
+    def get_export_symbols(self, ext):
+        initfunc_name = "PyInit_" + ext.name.split(".")[-1]
+        if initfunc_name not in ext.export_symbols:
+            ext.export_symbols.append(initfunc_name)
+        return ext.export_symbols
+
+    def build_extensions(self) -> None:
+        original_link_shared_object = self.compiler.link_shared_object
+
+        def gen_lib_options(compiler, library_dirs, runtime_library_dirs, libraries):
+            lib_opts = [compiler.library_dir_option(dir) for dir in library_dirs]
+
+            for dir in runtime_library_dirs:
+                lib_opts.extend(compiler.runtime_library_dir_option(dir))
+
+            for lib in libraries:
+                (lib_dir, lib_name) = os.path.split(lib)
+                if lib_dir:
+                    lib_file = compiler.find_library_file([lib_dir], lib_name)
+                    if lib_file:
+                        lib_opts.append(lib_file)
+                    else:
+                        compiler.warn(
+                            f"no library file corresponding to '{lib}' found (skipping)"
+                        )
+                else:
+                    lib_opts.append(compiler.library_option(lib))
+            return lib_opts
+
+        def _link_shared_object(
+                objects, output_filename, output_dir = None, libraries = None, library_dirs = None, runtime_library_dirs = None,
+                export_symbols = None, debug = None, extra_preargs = None, extra_postargs = None, build_temp = None, target_lang = None):
+            if not self.compiler.initialized:
+                self.compiler.initialize()
+
+            expected_sym = f'PyInit_{os.path.splitext(os.path.basename(output_filename))[0]}'
+            assert export_symbols and expected_sym in export_symbols, \
+                f"Expected {expected_sym!r} in export_symbols for {output_filename!r}, got {export_symbols!r}"
+            library_target = os.path.abspath(output_filename if output_dir is None else os.path.join(output_dir, output_filename))
+            output_dir = os.path.dirname(library_target)
+            objs = [f'{o.replace(":", "$:")}' for o in objects]
+            libraries, library_dirs, runtime_library_dirs = self.compiler._fix_lib_args(libraries, library_dirs, runtime_library_dirs)
+
+            with open(os.path.join(self.build_temp, 'build.ninja'), 'a') as f:
+                library_dirs = [f'"{lib}"' for lib in library_dirs]
+                runtime_library_dirs = [f'"{lib}"' for lib in runtime_library_dirs]
+
+                lib_opts = gen_lib_options(self.compiler, library_dirs, runtime_library_dirs, libraries)
+                export_opts = [f'/EXPORT:{sym}' for sym in (export_symbols or [])]
+
+                ld_args = (
+                    lib_opts + export_opts
+                )
+
+                if export_symbols is not None:
+                    (dll_name, dll_ext) = os.path.splitext(
+                        os.path.basename(output_filename)
+                    )
+                    implib_file = os.path.abspath(os.path.join(self.build_temp, self.compiler.library_filename(dll_name)))
+                    ld_args.append(f'/IMPLIB:"{implib_file}"')
+
+                f.write('\n'.join([
+                    'rule link',
+                    f'  command = "{self.compiler.linker}" /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO {" ".join(ld_args)} /out:$out @$out.rsp',
+                    '  rspfile = $out.rsp',
+                    '  rspfile_content = $in_newline',
+                    '',
+                    f'build {library_target.replace(":", "$:")}: link {" ".join(objs)}',
+                ])+'\n')
+
+            self.compiler.mkpath(output_dir)
+            subprocess.check_call(['ninja', '-C', os.path.abspath(self.build_temp), library_target])
+
+        if IS_WINDOWS:
+            self.compiler.link_shared_object = _link_shared_object
+
+        try:
+            super().build_extensions()
+        finally:
+            if IS_WINDOWS:
+                self.compiler.link_shared_object = original_link_shared_object
+
 setup(
     name=PACKAGE_NAME,
     version=get_package_version(),
